### ISSUE TITLE
parse5 emits namespaceURIs for unknown namespaces

### ISSUE THREAD

**Issue by Sebmaster** [inikulin/parse5#79]
It's not really a bug in parse5 per sé, but I'm not sure the behaviour as-is has any advantages:

Currently parse5 emits the full namespace uri of an element with a [namespaceURI != SVG, HTML, MATHML](https://github.com/inikulin/parse5/blob/master/lib/serializer/index.js#L91), which results in serialized code like this:
```html
<!DOCTYPE html>
<html xmlns:html="http://www.w3.org/XML/1998/namespace">
<http://www.w3.org/XML/1998/namespace:html:head attrib="abc">ac</http://www.w3.org/XML/1998/namespace:html:head>
<!-- comment -->
</html>
```

I'm not sure that's ever necessary - the prefix is already included in the tagName parse5 retrieves from the document adapter and the namespaces themselves are defined in attributes. Any reason for that being there or could that be changed?

**Comment by inikulin** [comment]
> Currently parse5 emits the full namespace uri of an element with a namespaceURI != SVG, HTML, MATHML

This is per [spec](https://html.spec.whatwg.org/multipage/syntax.html#serialising-html-fragments):

> If current node is an element in the HTML namespace, the MathML namespace, or the SVG namespace, then let tagname be current node's local name. Otherwise, let tagname be current node's qualified name.

However, provided markup looks odd. As far as I remember, parser will produce elements only in HTML, SVG or MathML namespaces. Any other namespace can be set via DOM only. Can you provide the source code that resulted in the given markup? BTW, how browsers behave?

**Comment by inikulin** [comment]
Oh, I've got it. Seems like you are implementing XML parsing for the jsdom with the SAX. And trying to serialize nodes with the parse5 serializer. Right? If so, the issue is that XML-fragment serialization differs from the HTML serialization and described in the separate normative section of the spec:
https://html.spec.whatwg.org/multipage/xhtml.html#serialising-xhtml-fragments

**Comment by Sebmaster** [comment]
> tagname be current node's local name. Otherwise, let tagname be current node's qualified name

That should be localName || prefix + ":" + localName (== tagName) then, not namespaceURI + ":" + tagName.

> Oh, I've got it. Seems like you are implementing XML parsing for the jsdom with the SAX. And trying to serialize nodes with the parse5 serializer. Right?

Right, but I think we could repro the same behaviour with the NS DOM methods.

**Comment by inikulin** [comment]
> That should be localName || prefix + ":" + localName (== tagName) then, not namespaceURI + ":" + tagName.

Right, thank you for pointing this out. But `Node.prefix` is the part of the DOM spec and should be calculated on the DOM level and has nothing to do with AST. I wonder how we can handle this gracefully. Maybe implement `getNodePrefix()` in the tree adapter which will return empty string in the current implementations, while jsdom's adapter will return actual `Node.prefx`?

**Comment by Sebmaster** [comment]
I think that shouldn't even be necessary, since tagName already contains the prefix it should just be a matter of never emitting the namespaceURI (at least if I'm not forgetting some special case).

**Comment by inikulin** [comment]
Makes sense, thanks. Is this urgent and require fix in 1.x or you can wait till 2.0?

**Comment by Sebmaster** [comment]
A patch release on 1.x would be nice since we'll need to major on 2.0 too, it's not urgent though.

**Comment by inikulin** [comment]
Published as v1.5.1

**Comment by Sebmaster** [comment]
:bow:
